File: src/platform/python/mgba/log.py
# Copyright (c) 2013-2016 Jeffrey Pfau
#
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.
from ._pylib import ffi, lib  # pylint: disable=no-name-in-module
from . import create_callback

create_callback("mLoggerPy", "log", "_pyLog")


def install_default(logger):
    Logger.install_default(logger)


def silence():
    Logger.install_default(NullLogger())


class Logger(object):
    FATAL = lib.mLOG_FATAL
    DEBUG = lib.mLOG_DEBUG
    INFO = lib.mLOG_INFO
    WARN = lib.mLOG_WARN
    ERROR = lib.mLOG_ERROR
    STUB = lib.mLOG_STUB
    GAME_ERROR = lib.mLOG_GAME_ERROR

    _DEFAULT_LOGGER = None

    def __init__(self):
        self._handle = ffi.new_handle(self)
        self._native = ffi.gc(lib.mLoggerPythonCreate(self._handle), lib.mLoggerPythonDestroy)

    @staticmethod
    def category_name(category):
        return ffi.string(lib.mLogCategoryName(category)).decode('UTF-8')

    @classmethod
    def install_default(cls, logger):
        cls._DEFAULT_LOGGER = logger
        lib.mLogSetDefaultLogger(logger._native)

    def log(self, category, level, message):
        print("{}: {}".format(self.category_name(category), message))

    def set_default_levels(self, levels):
        lib.mLoggerPythonFilterCreate(self._native, levels)

    def set_category_levels(self, category, levels):
        if isinstance(category, str):
            category = category.encode("UTF-8")
        lib.mLoggerPythonFilterSet(self._native, category, levels)

    def clear_filter(self):
        lib.mLoggerPythonFilterClear(self._native)


class NullLogger(Logger):
    def log(self, category, level, message):
        pass

File: src/platform/python/log.c
/* Copyright (c) 2013-2016 Jeffrey Pfau
 *
 * This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
#include "log.h"

static struct mStandardLogger _stdLogger;
static bool _stdLoggerReady = false;

static void _ensureStdLogger(void) {
	if (_stdLoggerReady) {
		return;
	}
	mStandardLoggerInit(&_stdLogger);
	_stdLogger.logToStdout = true;
	_stdLogger.logFile = NULL;
	_stdLoggerReady = true;
}

static void _pyLogShim(struct mLogger* logger, int category, enum mLogLevel level, const char* format, va_list args) {
	struct mLoggerPy* pylogger = (struct mLoggerPy*) logger;
	char message[256] = {0};
	vsnprintf(message, sizeof(message) - 1, format, args);
	_pyLog(pylogger, category, level, message);
}

struct mLogger* mLoggerPythonCreate(void* pyobj) {
	struct mLoggerPy* logger = malloc(sizeof(*logger));
	logger->d.log = _pyLogShim;
	logger->d.filter = NULL;
	logger->pyobj = pyobj;
	return &logger->d;
}

void mLoggerPythonDestroy(struct mLogger* logger) {
	mLoggerPythonFilterClear(logger);
	free(logger);
}

void mLoggerPythonFilterCreate(struct mLogger* logger, int defaultLevels) {
	if (!logger->filter) {
		logger->filter = malloc(sizeof(*logger->filter));
		mLogFilterInit(logger->filter);
	}
	logger->filter->defaultLevels = defaultLevels;
}

void mLoggerPythonFilterSet(struct mLogger* logger, const char* category, int levels) {
	if (!logger->filter) {
		mLoggerPythonFilterCreate(logger, mLOG_ALL);
	}
	mLogFilterSet(logger->filter, category, levels);
}

void mLoggerPythonFilterClear(struct mLogger* logger) {
	if (!logger || !logger->filter) {
		return;
	}
	mLogFilterDeinit(logger->filter);
	free(logger->filter);
	logger->filter = NULL;
}

void mLoggerPythonUseStdLogger(int defaultLevels) {
	_ensureStdLogger();
	_stdLogger.d.filter->defaultLevels = defaultLevels;
	mLogSetDefaultLogger(&_stdLogger.d);
}

void mLoggerPythonStdFilterSet(const char* category, int levels) {
	_ensureStdLogger();
	mLogFilterSet(_stdLogger.d.filter, category, levels);
}

void mLoggerPythonClearStdLogger(void) {
	mLogSetDefaultLogger(NULL);
	if (!_stdLoggerReady) {
		return;
	}
	mStandardLoggerDeinit(&_stdLogger);
	_stdLoggerReady = false;
}
